lib/phoenix/socket/transport.ex

```diff
@@ -388,6 +388,45 @@ defmodule Phoenix.Socket.Transport do
     end
   end
 
+  @doc false
+  def connect_info(conn, opts) do
+    keys = Keyword.get(opts, :connect_info, [])
+    
+    for key <- keys, into: %{} do
+      case key do
+        :peer_data ->
+          {:peer_data, Plug.Conn.get_peer_data(conn)}
+
+        :x_headers ->
+          {:x_headers, fetch_x_headers(conn)}
+
+        :uri ->
+          {:uri, fetch_uri(conn)}
+
+        _ ->
+          raise ArgumentError, "connection info keys are expected to be one of [:peer_data, :x_headers, :uri], got: #{inspect(key)}"
+      end
+    end
+  end
+
+  defp fetch_x_headers(conn) do
+    for {header, _} = pair <- conn.req_headers,
+        String.starts_with?(header, "x-"),
+        do: pair,
+        into: %{}
```

Comment:
I moved this to a comprehension but, now that i think about it, since header names may be duplicated, maybe we should return a list here? I.e. remove the `into: %{}` clause?

Reply:
yeah, you're right